### ISSUE TITLE
Add os.statx() function

### ISSUE THREAD

**Issue by 4712af1a-97b2-4201-a079-020c3db0175a** [python/cpython#83714]
BPO | [39533](https://bugs.python.org/issue39533)
--- | :---
Nosy | @vstinner, @tiran, @alexander255, @scrool
PRs | <li>python/cpython#19125</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2020-02-03.04:26:02.819>
labels = ['type-feature', 'library', '3.9']
title = 'Use `statx(2)` system call on Linux for extended `os.stat` information'
updated_at = <Date 2020-05-17.19:40:44.290>
user = 'https://github.com/alexander255'
```

bugs.python.org fields:
```python
activity = <Date 2020-05-17.19:40:44.290>
actor = 'christian.heimes'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2020-02-03.04:26:02.819>
creator = 'ntninja'
dependencies = []
files = []
hgrepos = []
issue_num = 39533
keywords = ['patch']
message_count = 4.0
messages = ['361265', '369128', '369149', '369150']
nosy_count = 5.0
nosy_names = ['vstinner', 'christian.heimes', 'ntninja', 'slow franklin', 'scrool']
pr_nums = ['19125']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue39533'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-136334
* gh-139178
* gh-140162
* gh-140185
* gh-140216
* gh-140402
* gh-140407
* gh-140432
* gh-140446
* gh-140484
* gh-141043
* gh-141069

**Comment by ntninja** [comment]
I have rebased the associated PR for Python 3.12-dev and added backward-compatible support for new `statx` features introduced in Linux 5.8+. Is the `stat` fallback still needed?

Also: What about linking with the `statx` C library symbol on systems that have it? Should it use a weak reference or `dlopen` instead to ensure that the Python build will work with an older C library? Or should it only unconditionally use the `syscall` function (in which case all the header definitions will need to be duplicated as well or it would still require a newer system to build)?

**Comment by hydrargyrum** [comment]
I'm interested in this, is there some progress since?

**Comment by vstinner** [comment]
> I'm interested in this, is there some progress since?

The PR is outdated. Someone has to update it.

**Comment by jbosboom** [comment]
I'm interested in working on this.  I've posted [a draft PR](https://github.com/python/cpython/pull/136334) as earnest effort (in the sense of earnest money), but we should agree on the answers to these questions before serious review.

### Do we fall back to stat?

The previous PR gh-19125 did not fall back to stat if statx was unavailable.  A comment in this issue objected to this because statx was relatively new at the time.  statx was introduced in Linux 4.11, released April 2017.  The oldest currently-supported upstream LTS branch is 5.4, released November 2019.  If I am reading the [RHEL life cycle](https://access.redhat.com/support/policy/updates/errata/) and [RHEL release dates](https://access.redhat.com/articles/3078) correctly, the oldest supported release is RHEL8, based on kernel 4.18.  It is likely that all mainstream Linux distributions ship kernels supporting statx.  On the other hand, there are probably embedded environments stuck on ancient kernels; do their users want to run the newest Python?

A complicating factor is default-deny seccomp syscall filters that have not been updated to allow statx.  If they deny with ENOSYS and we fall back to stat, Python will keep working under those filters; if we don't fall back, Python will almost certainly fail during startup.

The code to fall back is simple enough, but testing the fallback path would require running on an old kernel or under a seccomp filter.

My draft PR has a fallback, but no tests for the fallback.

### How do we call statx?

In order of increasing reliance on the build environment, here are the ways we can call statx:

1. Hardcode the syscall numbers for architectures we care about, provide our own definition of struct statx and related flags, and call via syscall().  Build is self-sufficient.

1. Use syscall numbers from sys/syscall.h, definitions of struct stat and flags from linux/stat.h, provide our own definitions of flags newer than statx itself, call via syscall().  Build requires Linux kernel userspace API headers >= 4.11.  (New userspace API headers can be used on old kernels; they're distinct from the headers used to build kernel modules.)

1. Call using libc's statx wrapper (weakly linked), provide our own definitions of flags newer than statx itself.  Requires glibc version 2.28 (August 2018) at build time and run time (but can fall back to stat if runtime libc is old).

My draft PR is option 3.

### How do we indicate missing information?

We won't always get everything we ask for in struct statx, depending on the kernel, filesystem type, and filesystem instance (e.g., ext4 supports btime, but really old ext4 filesystems may not).  For btime specifically, st_birthtime has never been exposed on Linux, so it's tempting to set it to None when it isn't available.  Code that checks the platform before using st_birthtime and code using `getattr(st, 'st_birthtime', None)` will work, but code that relies on catching AttributeError may not expect st_birthtime to be None -- and there's an example of the latter in test_os.py.  So we have to fill it in, either with 0 or with whatever statx returned (not necessarily 0).

The basic stats (also in struct stat) are always initialized, but their bits may not be set in stx_mask.  linux/stat.h gives CIFS as an example that might clear STATX_UID/GID if they're being overridden client-side, though I haven't observed this.  I have observed a CIFS mount on Linux 6.15.4 that doesn't set STATX_ATIME.  We obviously can't set st_atime/uid/gid to None.

I think we should indicate to the application what fields are and aren't reliable, but I don't know how.  Exposing stx_mask isn't extensible to other operating systems.  We could make up our own mask and flags.  We could have a set-like member for tests of the form `'st_birthtime' in st.present`, though I don't know how to implement that performantly (and stat is likely hot in some programs).  We could introduce int/float subclasses with an extra 'present' member, but I'm sure there are programs that expect exactly int/float.  I don't like any of these options.

My draft PR does nothing here.

### Which mount ID do we want?

There are two flags to ask for mount IDs, STATX_MNT_ID and STATX_MNT_ID_UNIQUE.  The non-unique ID indexes `/proc/self/mountinfo` and is subject to reuse race conditions; the unique ID can be used with the statmount system call, which requires CAP_SYS_ADMIN and does not currently have a glibc wrapper.  Unfortunately they both initialize stx_mnt_id, so we can only get one or the other (unless we call statx twice).  Setting both flags is not an error, but I do not think the result is defined; my experiments have always obtained the unique ID.

My draft PR provides the non-unique ID because I think it will be more useful to general Python programs.

### How do we test the new members?

ntfs-3g and cifs allow setting btime via the `user.ntfs/cifs.creationtime` magic xattr.  I don't know offhand if ntfs-3g reports btime (being a FUSE filesystem).  cifs does, but that requires a server to be available.  We could also just create a file, wait, modify the file, and check its btime differs from its mtime and ctime.

We can test stx_mnt_id by looking for it in `/proc/self/mountinfo`.

Only btrfs and bcachefs currently support stx_subvol.  [bcachefs has an uncertain future](https://lwn.net/Articles/1027289), so we'd want btrfs.

Of the stx_attributes, only STATX_ATTR_MOUNT_ROOT is easily testable, by opening `/proc/self/mountinfo` and checking the flag is set on mount points.  This flag should also be independent of both filesystems.

My draft PR adds no new tests.

### Do we want to map attributes to st_flags?

The documentation for the os module claims st_flags "may be available on some Unix systems (such as Linux)", but as far as I can tell it has never been exposed on Linux.  The previous PR gh-19125 mapped both Windows and Linux attribute bits to st_flags.  Do we want that?  Do we want to map st_flags back to attribute bits?  (I don't know which systems currently have st_flags or how they are used.  Many of them are macOS-specific.  The os.chflags documentation says "Availability: Unix, not WASI" but it's not present on Linux.  BSD?)

My draft PR doesn't do this.

### What do we name stx_attributes?

My PR names it st_attributes, but maybe st_linux_attributes is safer.

### Future work

statx supports the AT_STATX_FORCE_SYNC and AT_STATX_DONT_SYNC flags to ask for up-to-date or cached information.  (AT_STATX_SYNC_AS_STAT is 0.)  I'd like to expose this as a sync=True/False/None argument to os.stat, but I need to see what other operating systems offer first.  In particular, the statx flags are hints; if cached info is not available, statx will hit the network/tape/etc.  I can imagine that other operating systems might have a flag to say "cached if possible, else fail, never block", which would have to be handled differently.

**Comment by ntninja** [comment]
@jbosboom: Have you looked at my PR before writing yours?

### `st_birthtime` may be None

I’ve addressed the “`st_birthtime` may be None” issue by only exposing the new information as `st_btime` and adding `st_birthtime` as legacy alias on BSD only, so existing code checking for the presence of `st_birthtime` should not be affected (but will continue to only detect btime on BSD). I’ve also copied the value of `st_ctime` on Windows to `st_btime` since file creation time is exposed as `st_ctime` there (which also continues to work and be inconsistent across platforms).

So my patch `st_btime` is always defined but may be `None` in some cases on Linux – this is fully backwards-compatible and solves all issues.

This should also solve all cases of missing information in practise and set at precedent for future fields.

### Extra fields

I didn’t do this, but they should always be exposed on Linux only, but are allowed to be `None`.

### Do we want to map attributes to st_flags?

I had implemented mapping these for Linux based on BSD semantics IIRC.

Perhaps I did something about `st_attributes` as well, but I cannot remember anymore…

**Comment by jbosboom** [comment]
> [@jbosboom](https://github.com/jbosboom): Have you looked at my PR before writing yours?

Yes, and it was sufficiently useful in writing my PR that in my PR comments I offered you Co-authored-by and/or blurb credit.  Do you want it?

> ### `st_birthtime` may be None
> 
> I’ve addressed the “`st_birthtime` may be None” issue by only exposing the new information as `st_btime` and adding `st_birthtime` as legacy alias on BSD only, so existing code checking for the presence of `st_birthtime` should not be affected (but will continue to only detect btime on BSD). I’ve also copied the value of `st_ctime` on Windows to `st_btime` since file creation time is exposed as `st_ctime` there (which also continues to work and be inconsistent across platforms).

3.12 added `st_birthtime` on Windows.  That doesn't prevent us from adding a `st_btime` that may be None, but it would be more churn for code that recently changed to standardize on `st_birthtime`, and it does not generalize to the other members (see directly below).

> This should also solve all cases of missing information in practise and set at precedent for future fields.

As I mentioned, statx can report that the basic stats are invalid.  I have observed this with atime on cifs and sys/stat.h suggests some filesystems may do it with uid/gid too.  Would you introduce new members for them that may be None?  (And if not, why would btime be a special case?)

> ### Extra fields
> 
> I didn’t do this, but they should always be exposed on Linux only, but are allowed to be `None`.

I agree.  I don't think my comment asked about this.

> ### Do we want to map attributes to st_flags?
> 
> I had implemented mapping these for Linux based on BSD semantics IIRC.

Yes.  I was asking if that was desirable, given that many flags seem macOS-specific.  I assume you are in favor, given your PR.

**Comment by ntninja** [comment]
> > [@jbosboom](https://github.com/jbosboom): Have you looked at my PR before writing yours?
> 
> Yes, and it was sufficiently useful in writing my PR that in my PR comments I offered you Co-authored-by and/or blurb credit. Do you want it?

Yes, please credit as “Erin of Yukis”. Thank you!

> 
> > ### `st_birthtime` may be None
> > I’ve addressed the “`st_birthtime` may be None” issue by only exposing the new information as `st_btime` and adding `st_birthtime` as legacy alias on BSD only, so existing code checking for the presence of `st_birthtime` should not be affected (but will continue to only detect btime on BSD). I’ve also copied the value of `st_ctime` on Windows to `st_btime` since file creation time is exposed as `st_ctime` there (which also continues to work and be inconsistent across platforms).
> 
> 3.12 added `st_birthtime` on Windows. That doesn't prevent us from adding a `st_btime` that may be None, but it would be more churn for code that recently changed to standardize on `st_birthtime`, and it does not generalize to the other members (see directly below).

I just liked `st_btime` more than `st_birthtime` due to consistancy with other fields, I guess. 😅 

(See next anser.)

> > This should also solve all cases of missing information in practise and set at precedent for future fields.
> 
> As I mentioned, statx can report that the basic stats are invalid. I have observed this with atime on cifs and sys/stat.h suggests some filesystems may do it with uid/gid too. Would you introduce new members for them that may be None? (And if not, why would btime be a special case?)

I had kinda remembered that BASIC_STATS fields are somehow “more reliable” than other fields but it seems I misread that…

As such, I now instead suggest adding a new kwarg named `guess_missing` (defaulting to `True`). Setting it to `False` allows *any* field to be `None` rather than the original type. This seems like the most pythonic solution to me given the circumstances. You’re right that adding aliases for *all* fields that may be `None` is highly unreasonable, yes.

That way we also do not need to figure out how to do platform-independent field availability coding.

> > ### Do we want to map attributes to st_flags?
> > I had implemented mapping these for Linux based on BSD semantics IIRC.
> 
> Yes. I was asking if that was desirable, given that many flags seem macOS-specific. I assume you are in favor, given your PR.

Correct, yes 🙂

**Comment by jbosboom** [comment]
> As such, I now instead suggest adding a new kwarg named `guess_missing` (defaulting to `True`). Setting it to `False` allows _any_ field to be `None` rather than the original type. This seems like the most pythonic solution to me given the circumstances.

I like this the best of the options I've considered so far.  An application aware that fields may be missing shouldn't care about the faked/defaulted values of those fields; if it wants to use them anyway, it doesn't need to know they're fake.  The only use case for deliberately using a fake field that I can think of is to display it to the user with an asterisk next to it.

On Windows, uid/gid are currently always 0 (called "dummy values") in the docs.  These would become None instead in this mode.  Though apparently files can have uid/gid if they're created through WSL (see [FILE_STAT_LX_INFORMATION](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_file_stat_lx_information)); I don't know if it makes sense to expose that.

**Comment by ntninja** [comment]
> > As such, I now instead suggest adding a new kwarg named `guess_missing` (defaulting to `True`). Setting it to `False` allows _any_ field to be `None` rather than the original type. This seems like the most pythonic solution to me given the circumstances.
> 
> I like this the best of the options I've considered so far. An application aware that fields may be missing shouldn't care about the faked/defaulted values of those fields; if it wants to use them anyway, it doesn't need to know they're fake. The only use case for deliberately using a fake field that I can think of is to display it to the user with an asterisk next to it.

I know you weren’t actually concerned but given this API they’d instead just expose `None` as `"–"` or something and that’d be even *less* confusing in terms of UX.

> On Windows, uid/gid are currently always 0 (called "dummy values") in the docs. These would become None instead in this mode. Though apparently files can have uid/gid if they're created through WSL (see [FILE_STAT_LX_INFORMATION](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_file_stat_lx_information)); I don't know if it makes sense to expose that.

That’s easy! Just write in the docs that WSL uid/gid are currently not exposed… done. (Don’t think anyone will ever care about it, TBH.)

**Comment by gpshead** [comment]
_Pulling my comment from the PR out up here as it is probably a better place for the API design discussion:_

How about not modifying `os.stat` behavior at all but instead introduce an `os.statx` API that uses the more modern system call when available and exposes its functionality - in particular the mask functionality so that people can call it and get only the specific data they want instead of a large everything-bagel os.stat result.

The `os` module is all about providing wrappers around the underlying platform APIs.

-- https://github.com/python/cpython/pull/19125#issuecomment-3239639029

**Comment by jbosboom** [comment]
I want to at least provide `st_birthtime` from `os.stat` so that cross-platform code can use it.  `st_birthtime` was why this issue was filed (looking at the second comment), and it's been available on Windows since 3.12.  On Linux, getting the birthtime requires calling `statx`.  I wouldn't expect `statx` for `(STATX_BASIC_STATS | STATX_BTIME)` to be measurably slower than calling `stat`, but as noted in the PR, [weird things have happened at the libc/kernel boundary](https://lwn.net/Articles/944214/).

That said, I'd be happy to add `os.statx` with `mask` and `sync=True/False/None` parameters and that can always return None for missing fields (or return the possibly-fake fields and the mask, but checking the mask in Python will be slower).  That's much cleaner than trying to change `os.stat` given that other OSes don't have the concept of missing fields or requesting fresh/cached data.

**Comment by gpshead** [comment]
it may be entirely reasonable to do that, i think mostly it's a matter of testing to find out if there's a meaningful performance hit.  

regardless having stat use statx when possible _and_ having a specific featureful statx api could both wind up making sense.  it sounds like the larger win you're chasing is the os.stat specific additional data one - that makes sense to me.

making os.stat use it presumably means pathlib.Path.stat gets it as well as a result which seems good.

**Comment by cmaloney** [comment]
👍 I like having a `os.statx` which exposes all the nice functionality/flags that has directly to users. Evaluate on different Linux / lib c versions (get the bots all green) with that I think is a nice definitely improvement.

Once have that will know how to filter for "platforms that have `statx`", and can evaluate performance of moving [`_Py_fstat_noraise`](https://github.com/python/cpython/blob/c7a097c02ed4df9af227736e8bdc7d51e891eb83/Python/fileutils.c#L1233-L1288) and/or [`posix_do_stat`](https://github.com/python/cpython/blob/c7a097c02ed4df9af227736e8bdc7d51e891eb83/Modules/posixmodule.c#L2796-L2859) to use `statx` under the hood (which would provide more stat members by default).

Splitting into the two to me decouples risks and gets small shippable chunks.

**Comment by ExplodingCabbage** [comment]
My two cents as a naive Python end user (who isn't really literate in C and so is too dumb to contribute much beyond an end user perspective here): the thing that to me personally seems overwhelmingly most important is for there to be a simple, cross-platform one-liner built into the standard library that I can use to get a file creation date on all (possible) platforms.

The question about how to do this at https://stackoverflow.com/q/237079/1709587 (to which my answer is the top answer) has 1.3 *million* views (and note that Stack Overflow avoids double-counting if the same IP views a question twice). Because there is currently no one-line answer, my answer is kind of long and complicated; let's say it takes 5 minutes to read and comprehend. In aggregate, then, that's 12 *years* of Python programmer life spent understanding how to get file creation dates. If we can make my answer there redundant for future Python versions and have a one-line standard library solution at the top of it, maybe over the next decade we can save a decade of programmer lifetime from being squandered on this.

Making this one-line solution be to use the `st_birthtime` attribute from `os.stat` makes sense to me given that it already supports Windows, though to be honest I really dislike Python's choice to make `st_birthtime` cross-platform in the first place and if that hadn't already been done I'd think the correct API here would be a new top-level function in the `os` module (that can *optionally* take an existing `stat` or `statx` result as a parameter, so that users who are calling `stat` for other reasons anyway and care about performance can avoid making Python call it again).

Why do I dislike the existing decision to make `st_birthtime` support Windows? Two reasons, both of which immediately occurred to me years ago when I first learned about the change:
* it's weird and confusing that a macOS-and-FreeBSD-specific name was used for the cross-platform API; it will cause users familiar with `st_birthtime` from contexts *outside* Python and who aren't developing specifically for Mac to quite reasonably assume Python's `st_birthtime` can't be relevant to them before they even read the docs about it, and it will mis-educate users who *aren't* already familiar with file attributes to assume `birthtime` really is something that exists on multiple platforms. (Those users will then be confused when one day they try to do file metadata stuff _outside_ Python and discover that the view of the world Python was presenting them with was a lie)
* it was immediately obvious to me the moment I learned about this change that one day you guys might face *precisely* the awkward situation that you now do - where in order to *finish* implementing `st_birthtime` as a cross-platform attribute, you have to make `os.stat` cease to be a wrapper for `stat` on Linux and make it do other stuff necessary for getting the file creation time.

Everything ends up confusing as a result; the Linux user wanting a file creation time has to call `os.stat` in order to do something under the hood that isn't really `stat`, and then access the macOS-specific file attribute `birthtime` to get an underlying file attribute that isn't really called `birthtime` and isn't really macOS-specific. Horrible! Making these names have one meaning in the underlying platform APIs and a related but *different* meaning in the world's most popular programming language is a recipe for making a whole generation of programmers confused about how this stuff works.

(The performance considerations would also be simpler if getting file creation time were done via a dedicated top-level function. You would not have to worry about programs that don't really need file creation time needlessly doing extra work to get it when they call `os.stat` for another purpose.)

But Python seems to have already started down this path, and I think having a *partially* cross-platform `st_birthtime` attribute would be even more confusing, so unless there's an appetite for deprecating `st_birthtime` on Windows and adding a new top-level function for cross-platform fetching of file creation times, I reluctantly think making `st_birthtime` support Linux via `statx` seems like the logical extension of the choice Python already made to have it support Windows.

**Comment by vstinner** [comment]
> How about not modifying os.stat behavior at all but instead introduce an os.statx API that uses the more modern system call when available and exposes its functionality

I also prefer this plan.

**Comment by jbosboom** [comment]
I updated [my PR](https://github.com/python/cpython/pull/136334) (note, not ntninja's PR).  It's still a draft because it needs tests, docs and performance measurements.  I also wrote [a small C++ program](https://gist.github.com/jbosboom/a6dc8def084d87ed128e8a5f74c270c2) to make `statx` calls with customized arguments that may be useful for comparison purposes.

I'd like feedback on the following design decisions I'm not satisfied with.  (Sorry to use so many words.)

---

statx is enabled by default in configure on Linux (including Android).  If the wrapper function declaration is not present at configure time, configure fails with a message to either update glibc or use `--without-linux-statx`.  At runtime, we expose `os.statx` unless:
- the weakly-linked wrapper function is missing: we have an old runtime libc
- a trial statx call with AT_STATX_DONT_SYNC fails with EINVAL: glibc is emulating statx via stat (which is fine for `os.stat` because it doesn't use the sync flags)
- the trial statx call fails with ENOSYS: glibc was configured with --enable-kernel >= 4.11 but the runtime kernel doesn't have statx (`os.stat` will fall back to stat in this case); this really shouldn't happen but if we're already making a call we might as well check it

We have to (?) decide at compile time whether `os.stat_result` has `st_birthtime`.  If statx is enabled at configure time but isn't available at runtime, or if the filesystem doesn't support btime (e.g., ext2), we fill `st_birthtime[_ns]` with 0.0 and 0.  CPython's tests fail if `st_birthtime` is present but None, and programs probably fail similarly.  ntninja's PR handled this by introducing a new `st_btime` member that can be None.  That would not scale to all members possibly being None, but now that we have `os.statx` and `os.stat` no longer has a mask argument, we could consider it.  It's still unfortunate churn given we recently adopted `st_birthtime` on Windows.

I posted on discuss.python.org regarding [minimum Linux kernel and glibc versions for CPython](https://discuss.python.org/t/minimum-supported-linux-kernel-and-glibc-versions/103429).  There's no clear agreement.

---

`os.statx` returns `os.stat_result`.  I want `os.statx` to be a drop-in replacement for `os.stat` when only a subset of the members are used and/or `sync=False` is acceptable, for improved performance.  The new statx-specific members (other than `st_birthtime`) use the `stx_` prefix to help in explaining that `os.stat` leaves them None.  I don't love this -- it's confusing, it costs memory, and it costs some time to initialize the extra members to None.  Another option is to introduce a new `os.statx_result` and hope that duck-type-compatibility with `os.stat_result` is enough.  It isn't currently possible for a structseq to have bases.

---

By default, `os.statx` only initializes the members corresponding to `mask & stx_mask` (requested and valid) and the members without mask bits, leaving the rest None.  This is primarily for performance, both to avoid creating unused int and float objects and to avoid Python-level mask testing operations in favor of `is not None` (though `stx_mask` is always available, too).  I like the walrus operator, so I also think it provides better ergonomics:

```
if (btime := os.statx(path, os.STATX_BTIME).st_birthtime) is not None:
    # use btime here
```

versus

```
s = os.statx(path, os.STATX_BTIME)
if s.stx_mask & os.STATX_BTIME:
    # use s.st_birthtime here
```

But when called with `raw=True`, `os.statx` will fill in all members with the kernel-provided values.  In real programs, the only use case I can think of for deliberately using an invalid member is to display it to the user with an asterisk next to it, but `raw=True` is handy for investigating what the syscall does on different filesystems, and it provides an escape hatch for use cases I haven't foreseen.

What this doesn't provide is an option to initialize all valid members, even if they weren't requested in the mask.  Trying to "opportunistically" use members you didn't ask for just seems too complicated to be useful.

**Comment by cmaloney** [comment]
re: Members, I wonder if it would be possible to use something like https://docs.python.org/3/library/types.html#types.SimpleNamespace for `os.stat_result` which would make "member only sometimes" easier / work better (just add to the dictonary the members which are needed). Might be too much of a lift for the base "add os.statx" issue though. `os.stat_result` has been around for a long time (and its docs written a long time ago) so now newer mechanisms/tools exist in CPython which may make some parts simpler; but refactoring can take a lot of work (and validation of its own). 

Might also be a case where relying on duck-typing might give a path to use a newer tool for `os.statx` result which is compatible + virtual subclasses (https://docs.python.org/3/glossary.html#term-abstract-base-class) to make it look/flow through existing CPython code more easily.

**Comment by gpshead** [comment]
left comments on the PR.  TL;DR - os.statx should get a new modern return class namespacey type.  os.stat_result doesn't need fields that stat itself will never fill in.

**Comment by jbosboom** [comment]
@gpshead I've never implemented a Python type in C before, so can you please be more specific than "new modern return class namespacey type"? :)  I assume you mean not a tuple/structseq like `stat_result`, which is fine (modern code shouldn't be indexing or unpacking `stat_result`, so `statx_result` doesn't need that either).  But if you mean a type like `SimpleNamespace` where the set of attributes varies across instances, I think that turns off various interpreter optimizations.  I guess it still might be faster than a type with lots of fields initialized to None, so I'll look at it.

**Edit**: If you mean a regular extension type like in [tutorial section 2](https://docs.python.org/3/extending/newtypes_tutorial.html), I don't see how to give different instances different attributes or how to define class attributes for use as defaults.  If you mean to define a class in os.py with an `__init__` that takes kwargs, that would execute bytecode for each attribute, which seems too slow.  If you mean to define a class in os.py with no `__init__`, instantiate it from posixmodule.c, and set attributes directly, that could work.  I don't need to be hand-held here, but I do need some direction as to what to do.

I want to keep the `st_*` attribute names for duck-type compatibility with `stat_result`, so that code that only needs some of the fields can use `os.statx` for performance if available and `os.stat` otherwise.  Do y'all agree that is important?

If implementable and performant, I would also like to have unprefixed names, and to have major/minor separate like in `struct statx` rather than requiring the user to split them.  What do y'all think about that?

**Comment by cmaloney** [comment]
Re: Attribute names, 👍  on "all the `st_*`" names should work like before. I have a little bit of hesitation in that the "seconds" + "nanoseconds" variants is a bit complex, would be really nice to just have one which when accessed gets you a Python time object at resolution up to nanoseconds... I think it's more valuable that anywhere currently doing `os.stat()` can drop in `os.statx()` and code will keep just working which gives really nice refactoring paths for users.

re: Unprefixed names, I'm not a big fan as it makes a special "Translation to python names / idiom" vs. the C header names which other languages tend to match. If there was a performant (and simple to implement) way to get the `st_*{,_ns}` names + a python time object that helps avoid unit mistakes other than just int (the non-prefix non-suffix name) I'd maybe be for that. I don't see anything which seems to be well standardized there though (and a more normalized name / time system seems like something that would be really doable as a pypi module first)

Re: Implementation not sure exact CPython implementation pieces being referred to. https://discuss.python.org/t/add-builtins-namespace/78343/9 has some recent numbers around performance of some of the in-python ways to do it. Given this is all headed towards Python 3.15 and those numbers are 3.13 I suspect may be able to go even faster

**Comment by jbosboom** [comment]
> Re: Attribute names, 👍 on "all the `st_*`" names should work like before. I have a little bit of hesitation in that the "seconds" + "nanoseconds" variants is a bit complex, [...]

I expect most code to use only one or the other, so creating both a seconds-float and a nanoseconds-int does unnecessary work.  The only way I can think of to avoid that is to store the `struct statx` in the C `statx_result` object and use getters to materialize fields on-demand, which has its own cost.

**Comment by jbosboom** [comment]
I updated the PR to introduce a new `os.statx_result` type that stores members as attributes in its instance dict.  Unfortunately, performance is poor:

`python -m pyperf timeit -s 'import os' 'os.stat(".")'`: 1.16 us +- 0.01 us
`python -m pyperf timeit -s 'import os; m = os.STATX_BASIC_STATS | os.STATX_BTIME' 'os.statx(".", m)'`: 3.24 us +- 0.05 us
`python -m pyperf timeit -s 'import os' 'os.statx(".", 0)'`: 2.99 us +- 0.04 us

Now, @gpshead mentioned in a comment on the PR that `os.statx_result` should use its class attributes for default values.  The docs suggest that isn't possible for C heap types (see below), but it wouldn't help enough anyway.  Here are timings with the `SET_ATTR_IF` and `SET_ATTR_IF_BIT_TS` macros altered to not store the attribute when the bit is not set (instead of storing `None`):

`python -m pyperf timeit -s 'import os; m = os.STATX_BASIC_STATS | os.STATX_BTIME' 'os.statx(".", m)'`: 2.47 us +- 0.03 us
`python -m pyperf timeit -s 'import os' 'os.statx(".", 0)'`: 1.44 us +- 0.01 us

The last time is for allocating the `statx_result`, filling just the unconditional members (`stx_mask`, `st_blksize`, `stx_attributes` and `st_dev`), and freeing them.  It's slower than `os.stat` filling in everything.

Am I doing/not doing something that causes `os.statx_result` to be slow?

---

The docs suggest it isn't possible for a C heap type to have class attributes.  The [`tp_dict` docs](https://docs.python.org/3.15/c-api/typeobj.html#c.PyTypeObject.tp_dict) say:

> This field should normally be initialized to NULL before PyType_Ready is called; it may also be initialized to a dictionary containing initial attributes for the type. Once [PyType_Ready()](https://docs.python.org/3.15/c-api/type.html#c.PyType_Ready) has initialized the type, extra attributes for the type may be added to this dictionary only if they don’t correspond to overloaded operations (like [__add__()](https://docs.python.org/3.15/reference/datamodel.html#object.__add__)). Once initialization for the type has finished, this field should be treated as read-only.

along with a boxed warning not to modify tp_dict with the dictionary C API.

The [`PyType_Ready` docs](https://docs.python.org/3.15/c-api/type.html#c.PyType_Ready) say:

> This should be called on all type objects to finish their initialization.

So there isn't actually any time between `PyType_Ready` and `tp_dict` becoming read-only.  But we know the default values up front, so we should just specify them before calling `PyType_Ready`, right?  The [`PyType_Slot` docs](https://docs.python.org/3.15/c-api/type.html#c.PyType_Slot) say:

> The following internal fields cannot be set at all when creating a heap type: tp_dict, [...]

I suspect some part of these docs is wrong, but I don't know what.

---

I'm going to try an implementation using [PyGetSetDef](https://docs.python.org/3.15/c-api/structures.html#c.PyGetSetDef) to create getters that lazily unpack the `struct statx` as attributes are referenced.  That lets us avoid creating ints for the `struct statx` members that are unconditionally initialized (have no mask bit), and avoid creating both the seconds-float and nanoseconds-int timestamp representations in code that only uses one.

**Comment by jbosboom** [comment]
I think `os.statx_result` has to use `PyMemberDef` and/or `PyGetSetDef` to be competitive with `os.stat_result` (which is a structseq, which uses `PyMemberDef`).  There are a few options:

1. Create ints and floats for requested and valid (mask & stx_mask) members during `os.statx` and expose them using `PyMemberDef`.  Essentially `os.stat_result` with more members but not a tuple.
  - Pros: ints and floats created at most once; attribute access just loads the object by offset and increfs it; invalid fields can be tested with 'r.st_member is not None` as well as testing `stx_mask`.
  - Cons: some ints and floats created even if not used (the members with no mask bit, and both float and int timestamps when most code only uses one or the other).
2. Store `struct statx_result` in `os.statx_result`, plus C-level doubles and ints for timestamps, and expose members using `PyMemberDef`.
  - Pros: ints and floats only created when used.
  - Cons: ints and floats created multiple times if attribute is used multiple times; invalid fields exposed as is rather than replaced with None and can only be tested via `stx_mask`.
3. Store `struct statx_result` in `os.statx_result` and expose members via `PyGetSetDef` (only getters).
  - Pros: ints and floats only created when used; invalid fields can be tested with `r.st_member is not None` as well as testing `stx_mask`.
  - Cons: ints and floats created multiple times if attribute is used multiple times; getters are slightly slower than member descriptors.
4. Store `struct statx_result` in `os.statx_result` and expose members via `PyGetSetDef` (only getters), caching them in `os.statx_result` on first use.
  - Pros: ints and floats created at most once and only when used; invalid fields can be tested with `r.st_member is not None` as well as testing `stx_mask`.
  - Cons: `os.statx_result` larger than other options; getters are slightly slower than member descriptors; cache requires atomics in free-threaded build (member descriptors get them automatically).

We can mix and match some of these options, such as caching only some fields (basic stats + btime?) like option 4 and materializing the others from C-level ints like option 2.

---

I have some numbers from a hacky test that only has members/getters for `stx_mask` and `st_size`, but stores the full `struct statx` and cache members in the `statx_result` (just doesn't use most of them).

Here's the baseline of `os.stat`:
```
(venv) [jbosboom@promenade build]$ python -m pyperf timeit --rigorous -s 'import os' 'r = os.stat("."); r.st_size'
.........................................
Mean +- std dev: 1.19 us +- 0.01 us
(venv) [jbosboom@promenade build]$ python -m pyperf timeit --rigorous -s 'import os' 'r = os.stat("."); r.st_size; r.st_size; r.st_size; r.st_size'
.........................................
Mean +- std dev: 1.20 us +- 0.01 us
```

Here are numbers for option 2 showing the reuse cost:
```
(venv) [jbosboom@promenade build]$ python -m pyperf timeit --rigorous -s 'import os; m = os.STATX_BASIC_STATS | os.STATX_BTIME' 'r = os.statx(".", m)' 'if r.stx_mask & os.STATX_SIZE:' '  r.st_size'
.........................................
Mean +- std dev: 1.09 us +- 0.01 us
(venv) [jbosboom@promenade build]$ python -m pyperf timeit --rigorous -s 'import os; m = os.STATX_BASIC_STATS | os.STATX_BTIME' 'r = os.statx(".", m)' 'if r.stx_mask & os.STATX_SIZE:' '  r.st_size; r.st_size; r.st_size; r.st_size'
.........................................
Mean +- std dev: 1.16 us +- 0.01 us
```

Here's option 4 with space allocated to cache everything.  Note how `if not None` avoids materializing the mask int and checking it, and that repeated use is slightly cheaper but still not free:
```
(venv) [jbosboom@promenade build]$ python -m pyperf timeit --rigorous -s 'import os; m = os.STATX_BASIC_STATS | os.STATX_BTIME' 'r = os.statx(".", m)' 'if r.st_size is not None:' '  r.st_size'
.........................................
Mean +- std dev: 1.09 us +- 0.01 us
(venv) [jbosboom@promenade build]$ python -m pyperf timeit --rigorous -s 'import os; m = os.STATX_BASIC_STATS | os.STATX_BTIME' 'r = os.statx(".", m)' 'if r.stx_mask & os.STATX_SIZE:' '  r.st_size'
.........................................
Mean +- std dev: 1.12 us +- 0.01 us
(venv) [jbosboom@promenade build]$ python -m pyperf timeit --rigorous -s 'import os; m = os.STATX_BASIC_STATS | os.STATX_BTIME' 'r = os.statx(".", m)' 'if r.stx_mask & os.STATX_SIZE:' '  r.st_size; r.st_size; r.st_size; r.st_size'
.........................................
Mean +- std dev: 1.15 us +- 0.01 us
```

Allocating cache space for just the basic stats minus `st_blksize` and `st_blocks` plus `stx_mask`, `stx_attributes` and `stx_attributes_mask` (which I think are the "commonly used" members) isn't any faster given we only have one `statx_result`, but in a program that builds a dict of paths to stat results, we'd fill the CPU cache more slowly with the smaller structure.

Overall, there's surprisingly little room under `os.stat`.  Note that on a local disk filesystem (btrfs), there's no improvement to asking for just `STATX_SIZE`, at least compared to everything else Python is doing.  (I can't seem to get `perf` working to get time spent in the statx syscall specifically.)

**Comment by cmaloney** [comment]
On at least some systems CPython ships on the `stat` libc call is implemented by the libc with `statx` :). I don't have a strong preference among the options / My leaning would be benchmark / test / experiment and see what works best. To me cleanly exposing `statx` and its benefits (there's a reason it got added as a kernel API) is a really nice improvement

**Comment by gpshead** [comment]
fyi - tying threads together, in https://github.com/python/cpython/pull/136334#issuecomment-3303397390 and onwards we are advocating for a separate `os.statx` API - read there for why.

**Comment by serhiy-storchaka** [comment]
Since 3.3 it is preferable to extend existing functions instead of adding a new one for each syscall.

This would allow to add the birth time -- on FreeBSD it is returned by `stat()`, and on Linux by `statx()`.

**Comment by vstinner** [comment]
> Since 3.3 it is preferable to extend existing functions instead of adding a new one for each syscall.

In general yes. For the specific case of statx(), IMO it's too different from stat() and so requires its own os.statx() function.

See for example discussions about "sync" and "automount". There are subtle issues.

**Comment by serhiy-storchaka** [comment]
These issues already exist for `stat()` vs `statat()`. I think they were resolved in recent Linux kernels.

Why do we need `statx()` at first place? The original request was to provide fields for the birth time and the extra attributes. On FreeBSD they are returned by `stat()`, on Linux they are returned by `statx()`, on Windows -- by some other function. `os.stat()` is no longer a thin wrapper, it calls different underlying functions depending on platform and arguments. Adding `os.statx()` will not solve _this_ issue on FreeBSD and Windows. Open a new issue for adding separate `os.statx()`, but let first extend `os.stat()` for this issue.

**Comment by cmaloney** [comment]
Definitely can open a distinct issue for _just_ adding `os.statx`; agreed it's a little different shape than this initial issue.

I am not sure adding a new `os.stat` implementation in CPython itself which is backed by `statx` is valuable for the maintenance burden though as `libc` can already do that for performance (and some CPython platforms `statx` shows up like that), the `stat` codepath needs to be maintained with no clear deprecation coming across all supported platforms, and callers can't assume `st_btime` unless they are in the limited distribution area of only shipping on CPython versions with support and platforms with `statx`; in all other cases users need conditional fallback code.

Adding `os.statx` provides a path with less conditionals. It can be made available on older CPython versions fairly straightforward with a `hasattr` + "compat" / backport package. If a developer wants the extra `os.statx` attributes everywhere they can make that choice with a monkeypatch of `os.stat` and ship that for their codebase.

**Comment by serhiy-storchaka** [comment]
> Adding `os.statx` provides a path with less conditionals.

What to do on FreeBSD and Windows? Are you going to implement a limited version of `os.statx()` via `stat()` or `GetFileInformationByName()`? Or leave these platforms without a way to get the birth time even if the function used to implement `os.stat()` provides it?

**Comment by jbosboom** [comment]
@serhiy-storchaka The plan is apparently to have the [`Path.info` interface being discussed here](https://discuss.python.org/t/expose-file-size-mode-etc-from-pathlib-path-info/103828) use `os.statx` on Linux and `os.stat` everywhere else.  This was decided [in this comment on my previous PR](https://github.com/python/cpython/pull/136334#issuecomment-3307397806).

**Comment by vstinner** [comment]
For now, I would suggest focusing on adding `os.statx()`. Once the function lands in Python, we can discuss using it in `os.stat()`. So please review https://github.com/python/cpython/pull/139178 😄

**Comment by vstinner** [comment]
> For now, I would suggest focusing on adding os.statx(). Once the function lands in Python, we can discuss using it in os.stat(). So please review https://github.com/python/cpython/pull/139178 😄

I merged PR gh-139178. You can now play with `os.statx()` on the main branch. It gives access to many attributes which are not available on `os.stat()` (if you pass the right mask).

**Comment by vstinner** [comment]
Currently, `os.statx()` result set its member to `0` if they are missing from `stx_mask`. To use a `statx_result` attribute, you must check if it's set in `stx_mask`. 

```
$ ./python
>>> import os
>>> st=os.statx('.', 0)

>>> st.st_atime
1760626098.6986845
>>> st.st_mtime
0.0

>>> bool(st.stx_mask & os.STATX_ATIME)
True
>>> bool(st.stx_mask & os.STATX_MTIME)
False
```

For example, here `statx()` gives us `st_atime` for free, whereas `st_mtime` is not provided and set to `0`.

This behavior can lead to bug if the API is misused and IMO the API is surprising and error-prone.

I wrote https://github.com/python/cpython/pull/140216 to attempt a different approach: if the member is missing from `stx_mask`, set its value to None. The example becomes:

```
$ ./python
>>> import os
>>> st=os.statx('.', 0)

>>> print(repr(st.st_atime))
1760626235.2017012
>>> print(repr(st.st_mtime))
None
```

What do you think of this approach? Is it less error-prone and more "pythonic"? If yes, I will complete my draft PR (which currently only handle a few members).

**Comment by zooba** [comment]
FYI, there are systems out there (I have one) that define `STATX_MNT_ID` but don't have the `stx_mnt_id` field ([one example](https://patchwork.kernel.org/project/qemu-devel/patch/20220223092340.9043-1-hreitz@redhat.com/), [another example](https://github.com/util-linux/util-linux/issues/2415)).

I don't know what the logic here is, but it looks like people have used autoconf to detect the field properly, and not just relying on the flag.

**Comment by jbosboom** [comment]
As mentioned in [the documentation comments from `stat.h`](https://github.com/torvalds/linux/blob/98ac9cc4b4452ed7e714eddc8c90ac4ae5da1a09/include/uapi/linux/stat.h#L70), absent members aren't always 0:

```
 * For each bit in the mask argument:
 *
 * - if the datum is not supported:
 *
 *   - the bit will be cleared, and
 *
 *   - the datum will be set to an appropriate fabricated value if one is
 *     available (eg. CIFS can take a default uid and gid), otherwise
 *
 *   - the field will be cleared;
 * [...]
 * Items in STATX_BASIC_STATS may be marked unavailable on return, but they
 * will have values installed for compatibility purposes so that stat() and
 * co. can be emulated in userspace.
```

Similarly from the man page:

> If a filesystem does not support a field or if it has an unrepresentable value (for instance, a file with an exotic type), then the mask bit corresponding to that field will be cleared in stx_mask even if the user asked for it and a dummy value will be filled in for compatibility purposes if one is available (e.g., a dummy UID and GID may be specified to mount under some circumstances).

Besides the uid/gid mentioned, I've also observed that CIFS mounts provide atime when `STATX_ATIME` is in the mask, but don't set `STATX_ATIME` in `stx_mask`, I think because it's never guaranteed to be up to date with the server.

I assume the plan is still to use `os.statx` to implement `pathlib.Path.info`.  That interface will call `os.statx(self, os.STATX_BASIC_STATS | os.STATX_BTIME)`.  Maybe for btime pathlib will return None if not available, but for the basic stats pathlib will return whatever it gets, even if it's a fabricated value.  If we set missing members to `None` in `os.statx_result`, `Path.info` may have to fall back to `os.stat` to get the fabricated values.  (They'd still be fabricated; `os.statx` just admits when the kernel is lying.)

Possible options:

- We could treat `os.STATX_BASIC_STATS` specially and never set those members to None, even if they aren't set in `stx_mask`.  (Note that 0 can be a valid fabricated uid or gid, so checking for 0 doesn't help.)
- We could set missing members to None as you propose, but add a `raw` keyword argument to `os.statx` to return the raw values instead.  My first PR did this at one point, but I dropped it as being unnecessarily complex.

> IMO the API is surprising and error-prone.

With respect, I don't think the current behavior is surprising, because it directly matches the behavior of the system call.  We can't seem to decide whether `os.statx` should just wrap the system call or if it should provide a nice interface (e.g., relying on the user to pass `os.AT_NO_AUTOMOUNT` or providing `automount=False` default to match `os.stat` and family).

**Comment by jbosboom** [comment]
> FYI, there are systems out there (I have one) that define `STATX_MNT_ID` but don't have the `stx_mnt_id` field ([one example](https://patchwork.kernel.org/project/qemu-devel/patch/20220223092340.9043-1-hreitz@redhat.com/), [another example](https://github.com/util-linux/util-linux/issues/2415)).

Sigh.  From https://gitlab.com/qemu-project/qemu/-/issues/882, in which a user has the header with `struct statx.stx_mnt_id` but the compiler somehow doesn't find it:

>  Something seems broken, but I have no idea what it is.

I'll put together a PR to check for each field separately.

**Comment by vstinner** [comment]
> If a filesystem does not support a field or if it has an unrepresentable value (for instance, a file with an exotic type), then the mask bit corresponding to that field will be cleared in stx_mask even if the user asked for it and a dummy value will be filled in for compatibility purposes

For such case, I think that it's reasonable to use `None` and let the caller decides how to handle the case, no? It's not a big deal to ignore a dummy value. I mean `None` is not worse than a dummy value.

> Besides the uid/gid mentioned, I've also observed that CIFS mounts provide atime when STATX_ATIME is in the mask, but don't set STATX_ATIME in stx_mask, I think because it's never guaranteed to be up to date with the server.

My concern is that users passing `STATX_ATIME` to *mask* expect `statx_result.st_atime` to be set to a float without checking *stx_mask*. If Python provides the value `None`, it would force the caller to handle `None`. Well, it would say that it's more convenient that having to check *stx_mask*.

How are users supposed to handle this case?

```
>>> import os
>>> st=os.statx('.', 0)
>>> st.st_mtime
0.0
```

It's surprising to have a member available, but have a value of zero. In this case, mtime is missing from *stx_mask*:

```
>>> st.stx_mask & os.STATX_MTIME
0
```

> We could set missing members to None as you propose, but add a raw keyword argument to os.statx to return the raw values instead. My first PR did this at one point, but I dropped it as being unnecessarily complex.

I don't understand well in which case *raw* would be useful. Do you want to expose dummy values?

**Comment by jbosboom** [comment]
> ```
> >>> st=os.statx('.', 0)
> ```

Separately from the dummy values discussion, I see you keep calling `os.statx` with mask `0`, and earlier you were pushing for `0` to be the default.  I worry that this doesn't do what you think it does.  What do you think the mask parameter to `os.statx` is for?  What do you think a mask of `0` means?

---

> > We could set missing members to None as you propose, but add a raw keyword argument to os.statx to return the raw values instead. My first PR did this at one point, but I dropped it as being unnecessarily complex.
> 
> I don't understand well in which case _raw_ would be useful. Do you want to expose dummy values?

If you call `os.statx(path, os.STATX_BASIC_STATS)`, the fabricated/dummy values for the basic stats may be meaningful.  They're what `os.stat` would have returned.  That's documented in stat.h:
```
* Items in STATX_BASIC_STATS may be marked unavailable on return, but they
* will have values installed for compatibility purposes so that stat() and
* co. can be emulated in userspace.
```
We plan to use `os.statx` to implement `pathlib.Path.info` so it can provide btime when available.  `pathlib.Path.info` also wants to provide the basic stats.  If we expose the dummy values for the basic stats, `Path.info` can use them.  If we replace the dummy values for the basic stats with None, as you propose, `Path.info` will have to make an additional system call to `os.stat` to get those values.

I agree that checking for None is easier than checking `stx_mask`.  The problem is that the dummy values are sometimes useful.  If we do set missing members to None as you propose, the `raw` parameter is an escape hatch for the cases when the dummy values are useful.

**Comment by vstinner** [comment]
> We plan to use os.statx to implement pathlib.Path.info so it can provide btime when available

Oh ok, now I understand your use case: replace stat() with statx(). In that case, `statx(raw=True)` would make sense, I see.

I would prefer to stick to the bare minimum thin wrapper to the statx() function for now, and only consider adding *raw* parameter once we agree on a plan to use statx() in pathlib.

IMO it's better to use None for members missing from stx_mask. I see that it was discussed multiple times in previous discussions. I prefer to have a less error-prone API *by default*. As we wrote, we can extend the API later to add *raw*.

I plan to merge my PR once it will get a review.

**Comment by vstinner** [comment]
I merged my change gh-140216 to set `os.statx()` members to `None` if missing from `stx_mask`.

**Comment by encukou** [comment]
@vstinner This buildbot failure looks related: https://buildbot.python.org/#/builders/451/builds/6161

**Comment by vstinner** [comment]
> @vstinner This buildbot failure looks related: https://buildbot.python.org/#/builders/451/builds/6161

Oh. `st_birthtime` is None on files of /tmp which is mounted as a tmpfs even if STATX_BTIME mask was passed:

```
== cwd: /tmp/test_python_worker_903951æ
...
ERROR: test_statx_attributes (test.test_os.test_os.StatAttributeTests.test_statx_attributes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildbot/buildarea/3.x.cstratak-fedora-stable-ppc64le.clang-installed/build/target/lib/python3.15/test/test_os/test_os.py", line 812, in test_statx_attributes
    self.check_statx_attributes(self.fname)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
  File "/home/buildbot/buildarea/3.x.cstratak-fedora-stable-ppc64le.clang-installed/build/target/lib/python3.15/test/test_os/test_os.py", line 752, in check_statx_attributes
    self.check_timestamp_agreement(result, time_attributes)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/buildbot/buildarea/3.x.cstratak-fedora-stable-ppc64le.clang-installed/build/target/lib/python3.15/test/test_os/test_os.py", line 651, in check_timestamp_agreement
    floaty = int(getattr(result, name) * 100_000)
                 ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~
TypeError: unsupported operand type(s) for *: 'NoneType' and 'int'
```

Example:

```
$ python
>>> st=os.statx("/tmp", os.STATX_BTIME)
>>> print(repr(st.st_birthtime))
None
>>> st.stx_mask & os.STATX_BTIME
0
```

**Comment by jbosboom** [comment]
For the record, [tmpfs support for btime was added in Linux 5.18](https://github.com/torvalds/linux/commit/f7cd16a55837f37b4c3835a2c646023e4d0f0e04), released May 2022.

**Comment by cmaloney** [comment]
That buildbot is running `Linux-5.8.17-200.fc32.ppc64le-ppc64le-with-glibc2.41 little-endian`

**Comment by encukou** [comment]
Now we're getting other failures on CentOS & RHEL buildbots, e.g. https://buildbot.python.org/#/builders/1609/builds/3894 :
```
./Modules/posixmodule.c: In function ‘pystatx_result_get_stx_subvol’:
./Modules/posixmodule.c:3408:45: error: ‘STATX_SUBVOL’ undeclared (first use in this function)
 3408 | STATX_GET_ULONGLONG(stx_subvol, stx_subvol, STATX_SUBVOL)
      |                                             ^~~~~~~~~~~~
./Modules/posixmodule.c:3394:36: note: in definition of macro ‘STATX_GET_ULONGLONG’
 3394 |         if (!(self->stx.stx_mask & MASK)) { \
      |                                    ^~~~
./Modules/posixmodule.c:3408:45: note: each undeclared identifier is reported only once for each function it appears in
 3408 | STATX_GET_ULONGLONG(stx_subvol, stx_subvol, STATX_SUBVOL)
      |                                             ^~~~~~~~~~~~
./Modules/posixmodule.c:3394:36: note: in definition of macro ‘STATX_GET_ULONGLONG’
 3394 |         if (!(self->stx.stx_mask & MASK)) { \
      |         
```

**Comment by vstinner** [comment]
> ./Modules/posixmodule.c:3408:45: error: ‘STATX_SUBVOL’ undeclared (first use in this function)

Oh, it's hard to get this feature right :-(

**Comment by vstinner** [comment]
I ran a benchmark on Linux 6.16.3-200.fc42.x86_64 (Fedora 42):

```sh
$ env/bin/python -m pyperf timeit -s 'import os' 'os.stat(".")' -o stat.json -v
$ env/bin/python -m pyperf timeit -s 'import os; mask=os.STATX_BASIC_STATS' 'os.statx(".", mask)' -o statx.json -v
```

Result:

* btrfs (`/home`): `Mean +- std dev: [stat] 1.77 us +- 0.04 us -> [statx] 1.27 us +- 0.01 us: 1.39x faster`
* VFAT (`/boot/efi`): `Mean +- std dev: [stat] 1.65 us +- 0.03 us -> [statx] 1.35 us +- 0.01 us: 1.22x faster`
* ext4 (`/boot`): `Mean +- std dev: [stat] 1.78 us +- 0.04 us -> [statx] 1.25 us +- 0.01 us: 1.42x faster`

I'm surprised that `statx()` is faster than `stat()`, it's a good thing 😃

**Comment by jbosboom** [comment]
@vstinner `os.stat` eagerly creates Python objects for each of its attributes, while `os.statx` creates them lazily in getters.  A fairer comparison would access some of the attributes.  `os.statx` may still be faster because most users do not need all of the attributes.

**Comment by vstinner** [comment]
I changed the issue title to just add os.statx() function. I suggest creating a new issue if someone wants to modify os.stat() to use statx(), or use os.statx() in a higher level API such as pathlib.

Thanks everybody who was involved in this issue, first pull requests, reviews, tests, etc.

**Comment by cmaloney** [comment]
I was looking at buildbot logs and noticed a warning which seems to be originating from some of the `statx` code when built on 32 bit systems. If you look at https://buildbot.python.org/#/builders/1244/builds/6754 (Recent build of `main`/3.15 on x86 debian), there's these warnings:

```
./Modules/posixmodule.c:2638:31: warning: comparison is always true due to limited range of data type [-Wtype-limits]
./Modules/posixmodule.c:2638:45: warning: comparison is always true due to limited range of data type [-Wtype-limits]
```

Which seems to be coming from the the first bit of `stat_nanosecond_timestamp` added by this change:

https://github.com/python/cpython/blame/1ae900424b3c888d2b2cc97e6ef780717813d658/Modules/posixmodule.c#L2635-L2641

cc: @jbosboom , @vstinner

**Comment by vstinner** [comment]
I wrote https://github.com/python/cpython/pull/141043 to fix the two compiler warnings.
